src/libprojectM/Renderer/Renderer.cpp

```diff
@@ -81,36 +83,86 @@ Renderer::Renderer(int width, int height, int gx, int gy, int texsize, BeatDetec
 #endif /** USE_FTGL */
 
 
-	int size = (mesh.height - 1) *mesh.width * 5 * 2;
+    int size = (mesh.height - 1) *mesh.width * 4 * 2;
 	p = ( float * ) wipemalloc ( size * sizeof ( float ) );
 
 
 	for (int j = 0; j < mesh.height - 1; j++)
 	{
-		int base = j * mesh.width * 2 * 5;
+        int base = j * mesh.width * 4 * 2;
 
 
 		for (int i = 0; i < mesh.width; i++)
 		{
 			int index = j * mesh.width + i;
 			int index2 = (j + 1) * mesh.width + i;
 
-			int strip = base + i * 10;
-			p[strip + 2] = mesh.identity[index].x;
-			p[strip + 3] = mesh.identity[index].y;
-			p[strip + 4] = 0;
+            int strip = base + i * 8;
+            p[strip + 0] = mesh.identity[index].x;
+            p[strip + 1] = mesh.identity[index].y;
 
-			p[strip + 7] = mesh.identity[index2].x;
-			p[strip + 8] = mesh.identity[index2].y;
-			p[strip + 9] = 0;
+            p[strip + 4] = mesh.identity[index2].x;
+            p[strip + 5] = mesh.identity[index2].y;
 		}
 	}
 
+    renderContext.programID_v2f_c4f = shaderEngine.programID_v2f_c4f;
+    renderContext.programID_v2f_c4f_t2f = shaderEngine.programID_v2f_c4f_t2f;
 
 #ifdef USE_CG
 	shaderEngine.setParams(renderTarget->texsize, renderTarget->textureID[1], aspect, beatDetect, textureManager);
 #endif
 
+    // Interpolation VAO/VBO's
+    glGenBuffers(1, &m_vbo_Interpolation);
+    glGenVertexArrays(1, &m_vao_Interpolation);
+
+    glBindVertexArray(m_vao_Interpolation);
+    glBindBuffer(GL_ARRAY_BUFFER, m_vbo_Interpolation);
+
+    glEnableVertexAttribArray(0);
+    glVertexAttribPointer(0, 2, GL_FLOAT, GL_FALSE, sizeof(float)*4, (void*)0); // Positions
+
+    glDisableVertexAttribArray(1);
```

Comment:
i'm still learning this stuff so forgive my ignorance -
curious why this is index 1? after we just enabled index 0? and then the textures are index 2? what's index 1?

Reply:
The vertex shader (v2f_c4f_t2f_vert) used for interpolation takes 3 parameters:
```
layout(location = 0) in vec2 vertex_position
layout(location = 1) in vec4 vertex_color
layout(location = 2) in vec2 vertex_texture
```

`glEnableVertexAttribArray(0)` map an array of positions for the first parameter of the shader and `glVertexAttribPointer(0...` describe the array data format.

`glDisableVertexAttribArray(1)` allows to map a single color value for the second parameter of the shader, this color value is set by `glVertexAttrib4f(1, 1.0, 1.0, 1.0, pipeline.screenDecay)` in Renderer::Interpolation

The third parameter (textures coordinates), it is similar to the position array.